pkg/frontend/querymiddleware/astmapper/subtree_folder.go

```diff
@@ -65,48 +65,28 @@ func isVectorSelector(n parser.Node) (bool, error) {
 }
 
 // anyNode is a helper which walks the input node and returns true if any node in the subtree
-// returns true for the specified predicate function.
+// returns true for the specified predicate function. If fn returns an error, return that.
 func anyNode(node parser.Node, fn predicate) (bool, error) {
-	v := &visitor{
-		fn: fn,
+	result, err := fn(node)
+	if result || err != nil {
+		return result, err
 	}
 
-	if err := parser.Walk(v, node, nil); err != nil {
-		return false, err
+	for node := range parser.ChildrenIter(node) {
+		result, err := anyNode(node, fn)
+		if result || err != nil {
+			return result, err
+		}
 	}
-	return v.result, nil
+	return false, nil
 }
 
 // visitNode recursively traverse the node's subtree and call fn for each node encountered.
 func visitNode(node parser.Node, fn func(node parser.Node)) {
```

Comment:
Doesn't look like this is called in the `subtree_folder.go` file. Can we move it to the place it's called?